### ISSUE TITLE
Potential new projects to integrate with OSS-Fuzz

### ISSUE THREAD

**Issue by kcc** [google/oss-fuzz#402]
Let's use this  issue to maintain a list of projects that we want to see on oss-fuzz. 
Volunteers welcome! 

* [x]  [binutils](https://www.gnu.org/software/binutils/), e.g. `cplus_demangle` (seen fuzzable [bugs](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79099) in the past)
* [x]  more fuzz targets from [LLVM](http://llvm.org), we currently have only [llvm_libcxxabi](https://github.com/google/oss-fuzz/tree/master/projects/llvm_libcxxabi)
* [ ]  libstdc++ `regex` (e.g. https://gcc.gnu.org/bugzilla/show_bug.cgi?id=79539)
* [x]  [libexif](http://libexif.sourceforge.net/)
* [x]  [PCRE](http://pcre.org/) (we have PCRE2, but not PCRE)
* [x]  python (https://bugs.python.org/issue29505)
* [x]  gstreamer (https://lwn.net/Articles/708196/ and also https://github.com/google/oss-fuzz/issues/93)
* [x]  [mysql](https://www.mysql.com/)
* [x]  http://www.gfwx.org/
* [x]  http://www.libtiff.org/
* [x]  https://sourceforge.net/projects/giflib/
* [x]  https://github.com/jrmuizel/qcms/
* [x]  imagemagick @dlemstra (https://github.com/google/oss-fuzz/issues/430)
* [x]  ~~graphviz @emdenrg (https://github.com/google/oss-fuzz/issues/182)~~ ([upstream not interested](https://gitlab.com/graphviz/graphviz/-/issues/1370#note_285888414))
* [x]  gstreamer @bilboed (https://github.com/google/oss-fuzz/issues/93)
* [ ]  vim/neovim @dpelle (https://github.com/google/oss-fuzz/issues/96)
* [ ]  s2n @raycoll, @alexw91, @colmmacc (https://github.com/google/oss-fuzz/issues/238)
* [ ]  libelf, libdwarf (https://github.com/google/oss-fuzz/issues/64)
(add to this list instead of creating a new comment)
* [ ]  more/better fuzzing for ICU https://github.com/google/oss-fuzz/tree/master/projects/icu
* [x]  apache httpd
* [x]  https://xerces.apache.org/xerces-c/
* [x]  https://github.com/google/piex
* [x]  libidn2
* [x]  libzip
* [x] dovecot
* [x] Node.js
* [ ] tcpdump
* [ ] bluez
* [x] cairo
* [ ] coreboot
* [x] dbus
* [ ] dhcpcd
* [ ] dtc (device tree compiler)
* [ ] laptop-mode-tools
* [ ] libcamera
* [ ] libevdev
* [ ] libiio
* [ ] libpciaccess
* [ ] modemmanager
* [ ] xf86-input-synaptics
* [x] vlc
* [x] php-src
* [x] bind9
* [ ] projects from Cloud Native Computing Foundation(https://www.cncf.io/projects/)

**Comment by inferno-chromium** [comment]
[mysql](https://www.mysql.com/)

**Comment by kroeckx** [comment]
Some potential others:
- gzip
- bzip2
- xz
- lzma
- p7zip
- unzip
- libzip
- libarchive
(I think there are plenty of others like this.)
- tar
- cpio
- ar
- rpm
- dpkg
- elfutils
- ntp
- gimp
- inkscape
- cups
- ghostscript
- postfix
- dovecot
- libidn
- poppler
- clamav
- flac
- mutt
- vim
- libpcap
- privoxy

**Comment by bilboed** [comment]
Added GStreamer : https://github.com/google/oss-fuzz/pull/905

**Comment by Dor1s** [comment]
* VLC: https://github.com/videolan/vlc, https://www.videolan.org/security/

**Comment by inferno-chromium** [comment]
from Dor1s, 

There are few new fuzz targets in Chromium, need to add those (ideally, upstream all the targets); https://cs.chromium.org/chromium/src/third_party/zlib/contrib/tests/fuzzers/

**Comment by bshastry** [comment]
Hello,

I was wondering if adding the solidity compiler, the official "smart contract" compiler implementation, to oss-fuzz makes sense. See ethereum/solidity#5278

Thank you.

**Comment by Dor1s** [comment]
@bshastry, yes, most likely we would accept a PR for https://github.com/ethereum/solidity.

**Comment by bshastry** [comment]
> @bshastry, yes, most likely we would accept a PR for https://github.com/ethereum/solidity.

Thank you, I will initiate a PR once ethereum/solidity#5278 has been approved and merged into upstream solidity.

**Comment by bshastry** [comment]
Hello everyone,

Would you consider adding [mruby](https://github.com/mruby/mruby) to oss-fuzz? Hackerone features a [bounty program](https://hackerone.com/shopify-scripts) for memory corruption bugs in the mruby compiler

**Comment by Dor1s** [comment]
@bshastry, yes!

**Comment by fancycode** [comment]
> lzma

Would you still accept fuzzers for lzma? I recently started writing some for the LZMA SDK (https://www.7-zip.org/sdk.html) and would like to submit them here when they are ready.

**Comment by Dor1s** [comment]
> Would you still accept fuzzers for lzma?

Yes, definitely.

**Comment by bshastry** [comment]
Dear all,

What are top 5 projects you would like integrated into ossfuzz by the end of this year? I'm looking for work in my spare time, perhaps others following this thread are as well :-)

Thank you.

**Comment by DarkDimius** [comment]
Dear all, would you be interested in adding https://sorbet.org/ to oss-fuzz?
We already use fuzzing to find crashes in IDE before our users do, and we've found bugs through the entire typechecking pipeline via fuzzing before.

**Comment by inferno-chromium** [comment]
@DarkDimius - please submit a simple PR similar to https://github.com/google/oss-fuzz/pull/2536#issue-290927443, citing who is using Sorbet ide.

**Comment by catenacyber** [comment]
About elfutils (libelf, libdwarf see #64) clang compiler is currently not supported. (you can google clang elfutils for some references)
Is there a way to use another compiler with oss-fuzz ?

**Comment by Dor1s** [comment]
> Is there a way to use another compiler with oss-fuzz ?


No, it will not be feasible for us to support another toolchain. Even for existing LLVM toolchain, OSS-Fuzz actually relies on Chromium's judgement w.r.t. which revision is stable to use. We typically don't update it ourselves, but in the worst case we're capable of making some changes, which is not the case for other toolchains.

**Comment by catenacyber** [comment]
So, the prerequisite for integrating elfutils in oss-fuzz is to manage to build it with clang...

**Comment by Dor1s** [comment]
VLC keeps receiving memory corruption bugs: https://hackerone.com/vlc_h1c/hacktivity

**Comment by cmeister2** [comment]
First: @kcc, any chance your first post can be updated to remove projects that are now in OSS-Fuzz? This is a really useful list of potential projects for people to pick up.

Second: I've been trawling around for projects to add (I now have a lovely spreadsheet of projects and their functional areas), and I thought I'd list some here that might be useful in future.

- https://github.com/nothings/stb/ : Audio, Image, Font parsing.
  - There does appear to be a PR for an stb_image fuzzer: https://github.com/nothings/stb/pull/737
  - https://github.com/SFML/SFML uses the stb_image file; SFML is a fairly prevalent game development engine.
- https://pjsip.org: SIP (Internet Telephony)
  - PJSIP is a pretty well known SIP library, used in Asterisk.
  - I might take this on if I have the time, but the API is daunting 😅

**Comment by Dor1s** [comment]
> First: @kcc, any chance your first post can be updated to remove projects that are now in OSS-Fuzz? 

Updated the first comment on behalf of Kostya :)

**Comment by Dor1s** [comment]
Maybe Perl: https://hackerone.com/reports/424447

**Comment by practicalswift** [comment]
Bitcoin Core would be great to have fuzzed as part of oss-fuzz :)

I contribute to Bitcoin Core and would be very glad to help any volunteer who wants to give it a try!

https://github.com/bitcoin/bitcoin/blob/master/doc/fuzzing.md is a good starting point.

**Comment by bshastry** [comment]
https://github.com/sctplab/usrsctp ?

**Comment by bshastry** [comment]
libexpat has [one fuzzer](https://github.com/google/oss-fuzz/blob/master/projects/expat/parse_fuzzer.cc) in this repo and [two fuzzers](https://github.com/libexpat/libexpat/search?q=LLVMFuzzerTestOneInput&unscoped_q=LLVMFuzzerTestOneInput) in its upstream repo. Would you be interested in using the upstream version?

**Comment by inferno-chromium** [comment]
@bshastry - definitely, we can switch to upstream version, can delete current one in oss-fuzz repo, and see any of it needs to be upstreamed in either of those two fuzzers.

**Comment by catenacyber** [comment]
@bshastry there is this related open PR https://github.com/libexpat/libexpat/pull/292

**Comment by Dor1s** [comment]
> libexpat has one fuzzer in this repo 

also note that that one fuzz target is complied into six different binaries to fuzz different encoding formats

**Comment by bshastry** [comment]
@Dor1s upstream libexpat also does the same thing i.e., create 6 variants per harness, one per encoding type. In total, upstream contains two harnesses which means 12 fuzzers in total as against the 6 fuzzers in the oss fuzz repo

**Comment by abartlet** [comment]
Samba is in and working after #3094 and #2993

**Comment by bchurchill** [comment]
I'll work on elfutils, including compatibility with clang

**Comment by Dor1s** [comment]
Dovecot is another potential project to add, e.g. see https://hackerone.com/reports/546644

**Comment by Dor1s** [comment]
Node.js: https://hackerone.com/reports/340053

**Comment by Dor1s** [comment]
tcpdump: https://hackerone.com/reports/724253

**Comment by jonathanmetzman** [comment]
> tcpdump: https://hackerone.com/reports/724253

maybe we'll have luck poking https://github.com/google/oss-fuzz/pull/1757

**Comment by catenacyber** [comment]
> tcpdump: https://hackerone.com/reports/724253
> maybe we'll have luck poking #1757

tcpdump has a problem with the 90 days disclosure deadline
cf https://github.com/google/oss-fuzz/pull/1757#issuecomment-500755843

cc @fenner ? do you have more insight ?

But we can get rich if this is on hacker1 now ;-)

**Comment by fenner** [comment]
On Fri, Feb 14, 2020 at 2:05 PM Catena cyber <notifications@github.com>
wrote:

> tcpdump: https://hackerone.com/reports/724253
> maybe we'll have luck poking #1757
> <https://github.com/google/oss-fuzz/pull/1757>
>
> tcpdump has a problem with the 90 days disclosure deadline
> cf #1757 (comment)
> <https://github.com/google/oss-fuzz/pull/1757#issuecomment-500755843>
>
> cc @fenner <https://github.com/fenner> ? do you have more insight ?
>

Sorry, no.

  Bill

**Comment by sebras** [comment]
@kroeckx 
> Some potential others:
> * ghostscript

Ghostscript has been added.

**Comment by JulianVolodia** [comment]
Hello @johannkoenig and @fenner. With regards to your convo because it is not clear for me. Tcpdump *can't* be fuzzed because of maintenance issues like unrealistic 90d fix limit or nobody want to integrate it, for now?

@kcc can I ask you about integration of regex in libstdc++? We should integrate from exprerimental versions (worst cases), against some mainline package (from RHEL, Debian or Canonical) which deliver this or maybe, some bleeding edge and dont care about past, or any other answer? Thanks for your response.

**Comment by fenner** [comment]
On Tue, Mar 31, 2020 at 8:00 PM Volodia <notifications@github.com> wrote:

> Tcpdump *can't* be fuzzed because of maintenance issues like unrealistic
> 90d fix limit or nobody want to integrate it, for now?
>
My understanding from discussion with the tcpdump maintainers is that they
do not feel that the project has the resources to release a fix within 90
days of the bug's discovery, so would rather the bugs go un-discovered.

  Bill

**Comment by JulianVolodia** [comment]
Thanks @fenner. BTW. if I understand correctly, they (or other ppl) who will fix that maybe could have some support for their security related work because Google patch rewards.
Also, will it be against project and do You know some stories from the past where oss fuzz integration of any project end up with development collapse?

And, last but not least for You and other - will it be the problem if I integrate it anyway, but (idk of possibilities) fuzzing bug reports will be send out to them only in private? What you think? Or, am I missing something from FAQ or so?

**Comment by johannkoenig** [comment]
> Hello @johannkoenig and @fenner. With regards to your convo because it is not clear for me.

I'm not sure what you are referring to. I have not worked on tcpdump fuzzing.

**Comment by JulianVolodia** [comment]
> > Hello @johannkoenig and @fenner. With regards to your convo because it is not clear for me.
> 
> I'm not sure what you are referring to. I have not worked on tcpdump fuzzing.

hello @johannkoenig  - very sorry, I messed up things and mistakenly tagged You instead of jonathanmetzman. Forgive me. Have a good day or night.

**Comment by catenacyber** [comment]
@JulianVolodia there is an open PR for tcpdump integration here : https://github.com/google/oss-fuzz/pull/1757 with the specific discussion about it.

> And, last but not least for You and other - will it be the problem if I integrate it anyway, but (idk of possibilities) fuzzing bug reports will be send out to them only in private? What you think? Or, am I missing something from FAQ or so?

The tcpdump maintainers disagree for now and want to wait a bit longer

**Comment by DavidKorczynski** [comment]
Dovecot project added! https://github.com/google/oss-fuzz/pull/3502

**Comment by JulianVolodia** [comment]
Thanks @catenacyber for information with reference.

Can I know which project could be integrated next (== by me)? I had glance and see that many are still WIP or already prepared, but not referenced. Also, don't want to repeat story about TCPdump tool.

For example, other project from the list https://xerces.apache.org/xerces-c/ was initialy integrated and merged in PR #3083 . It is not 'checked' on the list above.

Because of Google's requirements from projects that can't be just random one or not prepared, I reviewed some PR and projects last day and wonder if I should look at repo and check each of them. For now it's unclear if there is some better way to pick any project to work on.

What you think? To not make more mess aka spam here and you could help me I will be very happy to read some rules of thumb (other than from docs which I read already). Thanks!

**Comment by catenacyber** [comment]
No silver bullet here.
If you do not have specific projects in mind, such as the tools you need during your work, the list above is the best thing I can think of...

But there is always the risk that someone is doing the same project already.
That happened to me for Dovecot and lzma SDK...
So, my best advice is to get a first PR quick, even if it is not complete/perfect, so that people know that you are working on it...

**Comment by JulianVolodia** [comment]
Thanks @catenacyber for your reply. I see, and understand. That is both sad (that you can't do it) and good (because oss-fuzz is alive and prosper) at one time. I hope you will get interesting projects soon. I wish best to you and will follow your advice.

**(0)** But, getting any project from repo which is integrated, but for example could have more fuzzers or sth will be the case here?
Ofc there is other reasons than improving security and gaining skills by practice, but I rather want to focus on that one mentioned for some time first.

**1)** Could I get any project outside of repo not choosen by Google to integrate in OSS fuzzing?

I started to check projects one by one here:  https://github.com/google/oss-fuzz/tree/master/projects if they have fuzzers (in their repo or here), and I assume only integrated are merged.

Upon the time, I going through items mentioned on 1st comment and PRs applying to them. Mention there this issue to show (like last mention about #3500 for `binutils, e.g. cplus_demangle (seen fuzzable bugs in the past)` ) item on list to be more easy to master the list.

Last but not least, **2)** Could improvement to integration of projects be the case here? In example new harness files.

I saw also, that there is some ideas to new ways of fuzzing out there, i.e. #1632 with data flow. Could I research on some?

Does fuzzing open source kernels would be the case to fuzz? Reference: https://github.com/fgsect/unicorefuzz
Some kernels out there: 
- Phoenix RT OS
 http://github.com/phoenix-rtos.com/phoenix-rtos-kernel
 http://phoenix-rtos.com/documentation/kernel
- SerenityOS
 https://github.com/SerenityOS/serenity
 http://serenityos.org/

I know that probably writing own harness and run it on own machine would be good but I highly believe in synergy and brainstorm before doing something. Sorry If my questions are dumb or too lame here.
Best regards, and sorry for long messages (maybe I should have open another thread/issue for that discussion). Thanks for reading.

**Comment by DavidKorczynski** [comment]
The cpp_demangle in binutils has been added (https://github.com/google/oss-fuzz/pull/3487). However, please follow here (https://github.com/google/oss-fuzz/issues/3500) since the project is not part of binutils as written above, but rather gcc.

**Comment by Dor1s** [comment]
@JulianVolodia please see some answers below:

> (0) But, getting any project from repo which is integrated, but for example could have more fuzzers or sth will be the case here?

yes, improving fuzzing coverage for projects that are already integrated is appreciated and rewarded (as an incremental ideal integration improvement)

> (1) Could I get any project outside of repo not choosen by Google to integrate in OSS fuzzing?

Yes, go ahead an upload a quick pull request with only project.yaml to find out whether the project of your choice will be accepted into OSS-Fuzz or not.

> Last but not least, (2) Could improvement to integration of projects be the case here? In example new harness files.

Yes!

> I saw also, that there is some ideas to new ways of fuzzing out there, i.e. #1632 with data flow. Could I research on some?

That's not so doable for an external contributor. Take a look at https://github.com/google/fuzzbench repo for ideas in this direction.

> Does fuzzing open source kernels would be the case to fuzz? Reference: https://github.com/fgsect/unicorefuzz

If it can be fuzzed in the user land, then it may be accepted into OSS-Fuzz. We don't do kernel fuzzing as of now.

**Comment by JulianVolodia** [comment]
Thank You @Dor1s for your descriptive answers! Really appreciate.

Btw. do You want some reminders/summaries if I do some issues brief read to forward summary update of whole, or it not wanted? Was said about abandon of entry "binutils cplus_demangle" in #3500. (which is not binutils but was written as it)

**Comment by Dor1s** [comment]
No need for any reminders / summaries. As you finish improving one project, go ahead and submit [this form](https://docs.google.com/forms/d/e/1FAIpQLSd5TlIXAiWRmbsHtPDR-8aDYKAZVgkJ5tcn6Dh-ym79r4iUxA/viewform) choosing OSS-Fuzz Ideal Integration and providing all the details about the work you've done.

Also make sure to check out https://security.googleblog.com/2017/05/oss-fuzz-five-months-later-and.html to better understand the rules and expectations.

**Comment by catenacyber** [comment]
What about openvpn ?
It is listed in this page https://www.google.com/about/appsecurity/patch-rewards/

**Comment by Dor1s** [comment]
> What about openvpn ?
> It is listed in this page https://www.google.com/about/appsecurity/patch-rewards/

@catenacyber I don't see a reason why not :)

**Comment by TheCrott** [comment]
If I found security bug in one of libraries integrated with oss-fuzz, am I eligible for bounty?

**Comment by inferno-chromium** [comment]
> If I found security bug in one of libraries integrated with oss-fuzz, am I eligible for bounty?

No, only if you add fuzzers that can find that security bug and other interesting ones. We want continuous fuzzing to find existing bugs and new ones.

> 
> > yes, improving fuzzing coverage for projects that are already integrated is appreciated and rewarded (as an incremental ideal integration improvement)
> 
> For "improving fuzzing coverage " Is it include adding more seeds?

Sure, we will consider this on a case-by-case basis. More significant coverage improvements are preferred.

**Comment by TheCrott** [comment]
After an integration finished, can I get bounty if I:
1. A contributor to that project
2. Not a contributor

**Comment by Dor1s** [comment]
@TheCrott yes in both cases

**Comment by TheCrott** [comment]
> @TheCrott yes in both cases

So how you distribute the bounty?

**Comment by JulianVolodia** [comment]
@TheCrott I asked about it some comments above. Thanks.

> No need for any reminders / summaries. As you finish improving one project, go ahead and submit [this form](https://docs.google.com/forms/d/e/1FAIpQLSd5TlIXAiWRmbsHtPDR-8aDYKAZVgkJ5tcn6Dh-ym79r4iUxA/viewform) choosing OSS-Fuzz Ideal Integration and providing all the details about the work you've done.

**Comment by ToSeven** [comment]
Hi everyone!  I have written a fuzz target for the libyuv, my code has been reviewed by the maintainer. https://github.com/google/oss-fuzz/pull/4363. If I modified my code and merged it to the OSS-Fuzz, Would I get a $1,000 bonus for the initial integration?

**Comment by Dor1s** [comment]
@ToSeven we can't promise anything here in the comments. The only way to find out is to make your contribution and submit the integration reward form after that for the OSS-Fuzz reward panel to review.

**Comment by harshkolhatkar** [comment]
Hey, sorry I'm new to this. Is there a document that I can refer to integrate OSS-Fuzz with the remaining projects?
